library/src/main/scala/scala/scalajs/concurrent/QueueExecutionContext.scala

```diff
@@ -1,22 +1,55 @@
 package scala.scalajs.concurrent
 
 import scala.concurrent.ExecutionContextExecutor
-import scalajs.js
 
-private[concurrent] object QueueExecutionContext
-    extends ExecutionContextExecutor {
+import scala.scalajs.js
+import scala.scalajs.js.|
 
-  def execute(runnable: Runnable): Unit = {
-    js.timers.setTimeout(0) {
-      try {
-        runnable.run()
-      } catch {
-        case t: Throwable => reportFailure(t)
-      }
+object QueueExecutionContext {
+  def timeouts(): ExecutionContextExecutor =
+    new TimeoutsExecutionContext
+
+  def promises(): ExecutionContextExecutor =
+    new PromisesExecutionContext
+
+  def apply(): ExecutionContextExecutor =
+    if (js.isUndefined(js.Dynamic.global.Promise)) timeouts()
+    else promises()
+
+  private final class TimeoutsExecutionContext extends ExecutionContextExecutor {
+    def execute(runnable: Runnable): Unit = {
+      js.Dynamic.global.setTimeout({ () =>
```

Comment:
Hum, there is no good reason to capture `setTimeout` here. The other guy "accidentally" captures promises because it wants to allocate exactly one promise for `()`. Not because it caches the `Promise` constructor per se.